### PR TITLE
feature: add new_form_creation_spec and creation_page

### PR DESCRIPTION
**Pull request by KirillGaliulin**
https://jira.railsc.ru/browse/AT-602

- Вынес проектный спек ПЦ https://github.com/abak-press/pulscen/blob/master/selenium/spec/product_creation_new_form_spec.rb в гем, чтобы проверки были и для БЛ
  - Перенес тесты, которые уже были
  - Добавил новые тесты
    - заполнение абсолютно всех полей на форме
    - модерация товара: отклонение, одобрение, отправка на рассмотрение
    - удаление товара
- Переписал спек и методы, которые были в спеке ПЦ
- Заскипал проверки для негативных кейсов загрузки фото, пока на ноже в БЛ не будет новый образ с файлами, которые грузятся в этих кейсах

Прогоны со своих форков
- БЛ http://allure-atm-bl2.knife.railsc.ru/index.html#
- ПЦ http://allure-atm-pc2.knife.railsc.ru/index.html

### CODE REVIEW

File: lib/pages/company_site/product_creation_page.rb [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+      end
+
+      last_mini_image_element.when_visible(30) if options[:qty_of_images]
+      self.description_of_photo = options[:description_of_photo] if options[:description_of_photo]
+    end
+
+    # Выбор группы
+    def set_group(options)
+      name_group = options[:group]
+      self.groups_tree_elements.find { |group| group.text.strip == name_group }.click
```
Comment (apress-hound) [inline lib/pages/company_site/product_creation_page.rb:253]: [Correctable] Style/RedundantSelf: Redundant self detected.

File: lib/pages/company_site/product_creation_page.rb [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+      else
+        upload_file(load_image_button_element, options[:path_to_image])
+      end
+
+      last_mini_image_element.when_visible(30) if options[:qty_of_images]
+      self.description_of_photo = options[:description_of_photo] if options[:description_of_photo]
+    end
+
+    # Выбор группы
+    def set_group(options)
```
Comment (apress-hound) [inline lib/pages/company_site/product_creation_page.rb:251]: Naming/AccessorMethodName: Do not prefix writer method names with set_.

File: lib/pages/company_site/product_creation_page.rb [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+    def set_company_traits(options)
+      options[:company_traits].each do |trait, value|
+        self.user_trait_input = trait
+        self.user_trait_value_input = value
+        add_user_trait
+      end
+    end
+
+    # Заполнение портальных характеристик
+    def set_portal_traits(options)
```
Comment (apress-hound) [inline lib/pages/company_site/product_creation_page.rb:217]: Naming/AccessorMethodName: Do not prefix writer method names with set_.

File: lib/pages/company_site/product_creation_page.rb [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+      finded_rubric
+    end
+
+    def set_priority_in_rubric(options)
+      options[:priority_lvl4] == true ? check_priority_lvl4_checkbox : uncheck_priority_lvl4_checkbox
+      options[:priority_lvl5] == true ? check_priority_lvl5_checkbox : uncheck_priority_lvl5_checkbox
+    end
+
+    # Заполнение компанейских характеристик
+    def set_company_traits(options)
```
Comment (apress-hound) [inline lib/pages/company_site/product_creation_page.rb:208]: Naming/AccessorMethodName: Do not prefix writer method names with set_.

File: lib/pages/company_site/product_creation_page.rb [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+
+    # Поиск и выбор рубрики
+    def set_rubric(options)
+      change_rubric_element.click unless change_rubric_not_exists? # редактирование рубрики, если есть кнопка
+      self.rubric_input = options[:rubric]
+      rubric_find_button
+      finded_rubric
+    end
+
+    def set_priority_in_rubric(options)
```
Comment (apress-hound) [inline lib/pages/company_site/product_creation_page.rb:202]: Naming/AccessorMethodName: Do not prefix writer method names with set_.

File: lib/pages/company_site/product_creation_page.rb [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+    end
+
+    # Заполнение полного описания
+    def description=(text)
+      wait_until { ckeditor? }
+      type_to_ck_editor('product_description', text) # метод из модуля apress-selenium_integration
+    end
+
+    # Поиск и выбор рубрики
+    def set_rubric(options)
```
Comment (apress-hound) [inline lib/pages/company_site/product_creation_page.rb:195]: Naming/AccessorMethodName: Do not prefix writer method names with set_.

File: lib/pages/company_site/product_creation_page.rb [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+      self.discount_price_old_input = options[:price_old]
+      self.discount_price_new_input = options[:price_new]
+      activate_calendar_discount_element.click
+      discount_without_expiration
+      self.currency_select = options[:currency] if options[:currency]
+      self.measure_select = options[:measure] if options[:measure]
+    end
+
+    # Заполнение оптовой цены
+    def set_wholesale_price(options)
```
Comment (apress-hound) [inline lib/pages/company_site/product_creation_page.rb:180]: Naming/AccessorMethodName: Do not prefix writer method names with set_.

File: lib/pages/company_site/product_creation_page.rb [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+    span(:thermometer_degree, css: '.aui-admin-battery__percent')
+    link(:thermometer_degree_popup_link, css: '.aui-admin-battery__link')
+    span(:thermometer_degree_popup, css: '.aui-admin-popup')
+
+    # ОПИСАНИЕ МЕТОДОВ
+    #
+    # Основной метод заполнения атрибутов, которые переданы в переменную options спека
+    # Если атрибут есть в переменной options, то заполняется методом
+    # Если атрибута нет, то пропускается
+    def fill_attributes(options)
```
Comment (apress-hound) [inline lib/pages/company_site/product_creation_page.rb:130]: Metrics/AbcSize: Assignment Branch Condition size for fill_attributes is too high. [<7, 46, 19> 50.26/35]<br>Metrics/CyclomaticComplexity: Cyclomatic complexity for fill_attributes is too high. [20/7]<br>Metrics/PerceivedComplexity: Perceived complexity for fill_attributes is too high. [20/8]

File: lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb [hunk 9]
```diff
@@ -0,0 +1,621 @@
...
+      end
+    end
+
+    context 'когда заполнены все основные поля' do
+      let(:options) do {
+        price: CONFIG['product_creation']['price']['exact'],
+        path_to_image: IMAGE_PATH,
+        short_description: CONFIG['product_creation']['short_description']['valid'],
+        description: CONFIG['product_creation']['full_description'],
+        exists: CONFIG['product_creation']['exists']['in_stock'] }
```
Comment (apress-hound) [inline lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb:359]: [Correctable] Layout/SpaceInsideHashLiteralBraces: Space inside } detected.<br>[Correctable] Layout/MultilineHashBraceLayout: Closing hash brace must be on the line after the last hash element when opening brace is on a separate line from the first hash element.

File: lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb [hunk 10]
```diff
@@ -0,0 +1,621 @@
...
+
+      before { @product_creation_page.exists_select = options[:exists] }
+
+      it "значение батарейки #{CONFIG['battery_percents']['exists']}%" do
+        expect(@product_creation_page.thermometer_degree).to eq "#{CONFIG['battery_percents']['exists']}%"
+      end
+    end
+
+    context 'когда заполнены все основные поля' do
+      let(:options) do {
```
Comment (apress-hound) [inline lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb:354]: [Correctable] Layout/MultilineBlockLayout: Block body expression is on the same line as the block start.

File: lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb [hunk 11]
```diff
@@ -0,0 +1,621 @@
...
+
+      before { @product_creation_page.description = options[:description] }
+
+      it "значение батарейки #{CONFIG['battery_percents']['description']}%" do
+        expect(@product_creation_page.thermometer_degree).to eq "#{CONFIG['battery_percents']['description']}%"
+      end
+    end
+
+    context 'когда заполнено Наличие' do
+      let(:options) { { exists: CONFIG['product_creation']['exists']['in_stock'] } }
```
Comment (apress-hound) [inline lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb:344]: [Correctable] Layout/SpaceInsideHashLiteralBraces: Space inside { detected.<br>[Correctable] Layout/SpaceInsideHashLiteralBraces: Space inside } detected.

File: lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb [hunk 12]
```diff
@@ -0,0 +1,621 @@
...
+
+      before { @product_creation_page.short_description = options[:short_description] }
+
+      it "значение батарейки #{CONFIG['battery_percents']['short_description']}%" do
+        expect(@product_creation_page.thermometer_degree).to eq "#{CONFIG['battery_percents']['short_description']}%"
+      end
+    end
+
+    context 'когда заполнено Полное описание' do
+      let(:options) { { description: CONFIG['product_creation']['full_description'] } }
```
Comment (apress-hound) [inline lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb:334]: [Correctable] Layout/SpaceInsideHashLiteralBraces: Space inside { detected.<br>[Correctable] Layout/SpaceInsideHashLiteralBraces: Space inside } detected.

File: lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb [hunk 13]
```diff
@@ -0,0 +1,621 @@
...
+
+      before { @product_creation_page.load_image(options) }
+
+      it "значение батарейки #{CONFIG['battery_percents']['image']}%" do
+        expect(@product_creation_page.thermometer_degree).to eq "#{CONFIG['battery_percents']['image']}%"
+      end
+    end
+
+    context 'когда заполнено Краткое описание' do
+      let(:options) { { short_description: CONFIG['product_creation']['short_description']['valid'] } }
```
Comment (apress-hound) [inline lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb:324]: [Correctable] Layout/SpaceInsideHashLiteralBraces: Space inside { detected.<br>[Correctable] Layout/SpaceInsideHashLiteralBraces: Space inside } detected.

File: lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb [hunk 14]
```diff
@@ -0,0 +1,621 @@
...
+
+          it 'под полем КО выводится предупреждение, что товар будет без КО, т.к. слова из названия' do
+            expect(@product_creation_page.short_description_error).to include '- слова из названия'
+          end
+        end
+
+        context 'когда введено > 50 символов и не совпадает с названием' do
+          let(:options) do {
+            name: CONFIG['product_creation']['name']['valid'],
+            short_description: CONFIG['product_creation']['short_description']['valid'] }
```
Comment (apress-hound) [inline lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb:250]: [Correctable] Layout/SpaceInsideHashLiteralBraces: Space inside } detected.<br>[Correctable] Layout/MultilineHashBraceLayout: Closing hash brace must be on the line after the last hash element when opening brace is on a separate line from the first hash element.

File: lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb [hunk 15]
```diff
@@ -0,0 +1,621 @@
...
+            @product_creation_page.short_description = options[:short_description]
+          end
+
+          it 'под полем КО выводится предупреждение, что товар будет без КО, т.к. слова из названия' do
+            expect(@product_creation_page.short_description_error).to include '- слова из названия'
+          end
+        end
+
+        context 'когда введено > 50 символов и не совпадает с названием' do
+          let(:options) do {
```
Comment (apress-hound) [inline lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb:248]: [Correctable] Layout/MultilineBlockLayout: Block body expression is on the same line as the block start.

File: lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb [hunk 16]
```diff
@@ -0,0 +1,621 @@
...
+
+          it 'под полем КО выводится предупреждение, что товар будет без КО, т.к. < 50 символов' do
+            expect(@product_creation_page.short_description_error).to include '- менее 50 символов'
+          end
+        end
+
+        context 'когда текст из КО совпадает с Названием' do
+          let(:options) do {
+            name: CONFIG['product_creation']['name']['valid'],
+            short_description: CONFIG['product_creation']['short_description']['as_the_name'] }
```
Comment (apress-hound) [inline lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb:234]: [Correctable] Layout/SpaceInsideHashLiteralBraces: Space inside } detected.<br>[Correctable] Layout/MultilineHashBraceLayout: Closing hash brace must be on the line after the last hash element when opening brace is on a separate line from the first hash element.

File: lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb [hunk 17]
```diff
@@ -0,0 +1,633 @@
...
+          @options[:min_qty] = Faker::Number.number(digits: 10)
+        end
+
+        @product_creation_page.fill_attributes(@options)
+        wait_until { @product_page.product_name? }
+      end
+
+      it 'товар сохранился с заполненными полями. Открывается КТ' do
+        expect(Page.browser.current_url).to match Regexp.new(
+          CONFIG['eti_company_subdomain'] + '.' + CONFIG['portal_page_url'] + CONFIG['url_intermediate_kt'] + '/')
```
Comment (apress-hound) [inline lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb:83]: [Correctable] Style/StringConcatenation: Prefer string interpolation to string concatenation.<br>[Correctable] Layout/MultilineMethodCallBraceLayout: Closing method call brace must be on the line after the last argument when opening brace is on a separate line from the first argument.

File: lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb [hunk 18]
```diff
@@ -0,0 +1,633 @@
...
+          description_of_photo: Faker::Lorem.paragraph_by_chars(number: 255),
+          group: CONFIG['product_creation']['group'],
+          path_to_doc: IMAGE_PATH,
+          qty_of_doc: 3, # на самом деле будет загружено 5 документов - это максимум
+          soputka: CONFIG['product_creation']['soputka'],
+          qty_of_soputka: 4
+        }
+
+        # На БЛ нет этого поля, только на ПЦ
+        if @product_creation_page.min_qty_input?
```
Comment (apress-hound) [inline lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb:73]: [Correctable] Style/IfUnlessModifier: Favor modifier if usage when having a single-line body. Another good alternative is the usage of control flow &&/||.

File: lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb [hunk 19]
```diff
@@ -0,0 +1,633 @@
...
+          rubric: CONFIG['rubrics_names_by_level'][5],
+          priority_lvl4: true,
+          priority_lvl5: true,
+          company_traits: {
+            CONFIG['product_creation']['company_traits']['traits'][1] => CONFIG['product_creation']['company_traits']['values'][1],
+            CONFIG['product_creation']['company_traits']['traits'][2] => CONFIG['product_creation']['company_traits']['values'][2],
+          },
+          portal_traits: {
+            CONFIG['product_creation']['portal_traits']['traits'][1] => CONFIG['product_creation']['portal_traits']['values'][1],
+            CONFIG['product_creation']['portal_traits']['traits'][2] => CONFIG['product_creation']['portal_traits']['values'][2],
```
Comment (apress-hound) [inline lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb:60]: Layout/LineLength: Line is too long. [129/120]

File: lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb [hunk 20]
```diff
@@ -0,0 +1,633 @@
...
+          tag_title: Faker::Lorem.characters(number: 255),
+          rubric: CONFIG['rubrics_names_by_level'][5],
+          priority_lvl4: true,
+          priority_lvl5: true,
+          company_traits: {
+            CONFIG['product_creation']['company_traits']['traits'][1] => CONFIG['product_creation']['company_traits']['values'][1],
+            CONFIG['product_creation']['company_traits']['traits'][2] => CONFIG['product_creation']['company_traits']['values'][2],
+          },
+          portal_traits: {
+            CONFIG['product_creation']['portal_traits']['traits'][1] => CONFIG['product_creation']['portal_traits']['values'][1],
```
Comment (apress-hound) [inline lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb:59]: Layout/LineLength: Line is too long. [129/120]

File: lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb [hunk 21]
```diff
@@ -0,0 +1,633 @@
...
+          wholesale_qty: Faker::Number.number(digits: 6),
+          short_description: Faker::Lorem.paragraph_by_chars(number: 255),
+          description: CONFIG['product_creation']['full_description'],
+          tag_title: Faker::Lorem.characters(number: 255),
+          rubric: CONFIG['rubrics_names_by_level'][5],
+          priority_lvl4: true,
+          priority_lvl5: true,
+          company_traits: {
+            CONFIG['product_creation']['company_traits']['traits'][1] => CONFIG['product_creation']['company_traits']['values'][1],
+            CONFIG['product_creation']['company_traits']['traits'][2] => CONFIG['product_creation']['company_traits']['values'][2],
```
Comment (apress-hound) [inline lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb:56]: Layout/LineLength: Line is too long. [131/120]

File: lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb [hunk 22]
```diff
@@ -0,0 +1,633 @@
...
+          wholesale_price: Faker::Number.number(digits: 10),
+          wholesale_qty: Faker::Number.number(digits: 6),
+          short_description: Faker::Lorem.paragraph_by_chars(number: 255),
+          description: CONFIG['product_creation']['full_description'],
+          tag_title: Faker::Lorem.characters(number: 255),
+          rubric: CONFIG['rubrics_names_by_level'][5],
+          priority_lvl4: true,
+          priority_lvl5: true,
+          company_traits: {
+            CONFIG['product_creation']['company_traits']['traits'][1] => CONFIG['product_creation']['company_traits']['values'][1],
```
Comment (apress-hound) [inline lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb:55]: Layout/LineLength: Line is too long. [131/120]

File: lib/pages/company_site/product_creation_page.rb [hunk 23]
```diff
@@ -0,0 +1,372 @@
...
+
+        clear_field
+      end
+    end
+
+    # Метод для перезагрузки страницы и ожидания отрисовки сообщения модерации
+    #   Очень редко может зафейлить. При быстрых действиях может отрисоваться некорректный статус
+    #   на странице и неверные кнопки модерации, поэтому ловим исключение и перезагружаем страницу
+    def wait_moderation_message
+      begin
```
Comment (apress-hound) [inline lib/pages/company_site/product_creation_page.rb:364]: [Correctable] Style/RedundantBegin: Redundant begin block detected.

File: lib/pages/company_site/product_creation_page.rb [hunk 24]
```diff
@@ -0,0 +1,372 @@
...
+    #   load_image(options) if options[:path_to_image]
+    #   fill_group(options) if options[:group]
+    #   load_doc(options) if options[:path_to_doc]
+    #   fill_soputka(options) if options[:soputka]
+    #   self.min_qty_input = options[:min_qty] if options[:min_qty]
+
+    #   save
+    # end
+
+    #
```
Comment (apress-hound) [inline lib/pages/company_site/product_creation_page.rb:159]: [Correctable] Layout/EmptyComment: Source code comment is empty.<br>[Correctable] Layout/TrailingWhitespace: Trailing whitespace detected.

File: lib/apress/selenium_products/version.rb [hunk 25]
```diff
@@ -2,6 +2,6 @@
 
 module Apress
   module SeleniumProducts
-    VERSION = '1.0.0'.freeze
+    VERSION = '1.0.1'.freeze
```
[the remark this answers is not in the document]
Comment (DenisDenis9331) [reply lib/apress/selenium_products/version.rb:5]: каждый раз будет ругаться, давай уберем `.freeze` так как уже фризим комментом `# frozen_string_literal: true`

[the remark this answers is not in the document]
Comment (KirillGaliulin) [reply lib/apress/selenium_products/version.rb]: поправил

File: lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb [hunk 26]
```diff
@@ -0,0 +1,635 @@
...
+            @product_creation_page.short_description = options[:short_description]
+          end
+
+          it 'под полем КО нет предупреждения' do
+            expect(@product_creation_page.short_description_error?).to be false
+          end
+        end
+      end
+
+      # TODO: расскипать (см. ниже)
```
Comment (DenisDenis9331) [inline lib/apress/selenium_products/spec/company_site/product_creation_page/new_form_spec.rb:276]: не увидел где смотреть)

**Review by DenisDenis9331 — APPROVED** [review]

**Review by yaros1avsv — APPROVED** [review]

**Comment by KirillGaliulin** [comment]
@DenisDenis9331 pls :)

**Comment by DenisDenis9331** [comment]
Конфликт получился, надо подлить мастер)

**Comment by KirillGaliulin** [comment]
Подлил мастер

**Pull request merged**